File: packages/client/src/file-crypto.ts
import {
  FILE_TRANSFER_PROTOCOL_VERSION,
  decodeFileFrame,
  encodeFileFrame,
  fileFrameAuthenticatedData,
  type FileFrameHeader,
  type FileFrameKind,
  type FileTransferDirection
} from "@mdbase-dev/connect-protocol";
import {
  RelayCryptoError,
  deriveP256SharedSecret,
  validateGrantEncryption,
  type GrantKeyStore,
  type RelayBinding
} from "./crypto.js";

const FILE_KEY_INFO = "mdbase-connect file chunk key v1";
const UUID_PATTERN =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;

export interface FileTransferBinding extends RelayBinding {
  authorityId: string;
  transferId: string;
  direction: FileTransferDirection;
}

export class FileTransferCryptoError extends Error {
  constructor(public readonly code: string, message: string) {
    super(message);
    this.name = "FileTransferCryptoError";
  }
}

/** Grant-bound, direction-separated cipher for one resumable file transfer. */
export class GrantFileTransferCipher {
  private constructor(
    private readonly binding: FileTransferBinding,
    private readonly key: CryptoKey
  ) {}

  static async open(
    store: GrantKeyStore,
    handle: string,
    binding: FileTransferBinding
  ): Promise<GrantFileTransferCipher> {
    validateFileTransferBinding(binding);
    const record = await store.get(handle);
    if (!record || record.agreementPublicKey !== binding.encryption.application_agreement_public_key) {
      throw new RelayCryptoError(
        "missing_grant_key",
        "The encrypted grant key is unavailable or does not match the file transfer grant."
      );
    }
    const sharedSecret = await deriveP256SharedSecret(
      record.agreementPrivateKey,
      binding.encryption.connector_agreement_public_key
    );
    return GrantFileTransferCipher.fromSharedSecret(sharedSecret, binding);
  }

  /** @internal Enables non-browser key-agreement adapters and compatibility tests. */
  static async fromSharedSecret(
    sharedSecret: BufferSource,
    binding: FileTransferBinding
  ): Promise<GrantFileTransferCipher> {
    validateFileTransferBinding(binding);
    const key = await deriveFileTransferKey(sharedSecret, binding);
    return new GrantFileTransferCipher(binding, key);
  }

  async encryptChunk(
    kind: FileFrameKind,
    header: FileFrameHeader,
    plaintext: Uint8Array
  ): Promise<Uint8Array> {
    this.validateHeader(header);
    if (plaintext.byteLength !== header.plaintext_length) {
      throw new FileTransferCryptoError(
        "plaintext_length_mismatch",
        "File chunk plaintext length does not match its authenticated header."
      );
    }
    const authenticatedData = fileFrameAuthenticatedData(kind, header);
    const payload = await crypto.subtle.encrypt({
      name: "AES-GCM",
      iv: chunkNonce(header.chunk_index),
      additionalData: arrayBuffer(authenticatedData),
      tagLength: 128
    }, this.key, arrayBuffer(plaintext));
    return encodeFileFrame({ kind, header, payload: new Uint8Array(payload) });
  }

  async decryptChunk(encoded: Uint8Array): Promise<Uint8Array> {
    const frame = decodeFileFrame(encoded);
    this.validateHeader(frame.header);
    const authenticatedData = fileFrameAuthenticatedData(frame.kind, frame.header);
    try {
      const plaintext = await crypto.subtle.decrypt({
        name: "AES-GCM",
        iv: chunkNonce(frame.header.chunk_index),
        additionalData: arrayBuffer(authenticatedData),
        tagLength: 128
      }, this.key, arrayBuffer(frame.payload));
      return new Uint8Array(plaintext);
    } catch {
      throw new FileTransferCryptoError(
        "authentication_failed",
        "The encrypted file chunk could not be authenticated."
      );
    }
  }

  private validateHeader(header: FileFrameHeader): void {
    const encryption = this.binding.encryption;
    if (header.protocol_version !== FILE_TRANSFER_PROTOCOL_VERSION
      || header.protection !== "grant_aead_v1"
      || header.grant_id !== this.binding.grantId
      || header.authority_id !== this.binding.authorityId
      || header.collection_id !== encryption.collection_id
      || header.transfer_id !== this.binding.transferId
      || header.direction !== this.binding.direction
      || header.scope_epoch !== encryption.scope_epoch
      || header.key_id !== encryption.key_id) {
      throw new FileTransferCryptoError(
        "header_binding_mismatch",
        "The file frame does not match its transfer encryption binding."
      );
    }
  }
}

/** @internal Deterministic HKDF profile shared with connector authorities. */
export async function deriveFileTransferKey(
  sharedSecret: BufferSource,
  binding: FileTransferBinding
): Promise<CryptoKey> {
  const contextBytes = new TextEncoder().encode(fileTransferContext(binding));
  const salt = await crypto.subtle.digest("SHA-256", contextBytes);
  const material = await crypto.subtle.importKey("raw", sharedSecret, "HKDF", false, ["deriveKey"]);
  return crypto.subtle.deriveKey({
    name: "HKDF",
    hash: "SHA-256",
    salt,
    info: new TextEncoder().encode(FILE_KEY_INFO)
  }, material, { name: "AES-GCM", length: 256 }, false, ["encrypt", "decrypt"]);
}

function validateFileTransferBinding(binding: FileTransferBinding): void {
  validateGrantEncryption(binding.encryption);
  if (!UUID_PATTERN.test(binding.grantId)
    || !UUID_PATTERN.test(binding.applicationId)
    || !UUID_PATTERN.test(binding.authorityId)
    || !UUID_PATTERN.test(binding.transferId)
    || (binding.direction !== "upload" && binding.direction !== "download")) {
    throw new FileTransferCryptoError(
      "invalid_binding",
      "The file transfer encryption binding is invalid."
    );
  }
}

function fileTransferContext(binding: FileTransferBinding): string {
  const encryption = binding.encryption;
  return [
    "mdbase-connect",
    "file-transfer",
    FILE_TRANSFER_PROTOCOL_VERSION,
    encryption.suite,
    binding.grantId,
    binding.applicationId,
    encryption.connector_id,
    binding.authorityId,
    encryption.collection_id,
    encryption.scope_epoch,
    encryption.key_id,
    binding.transferId,
    binding.direction
  ].join("|");
}

function chunkNonce(index: number): Uint8Array<ArrayBuffer> {
  if (!Number.isSafeInteger(index) || index < 0) {
    throw new FileTransferCryptoError("invalid_chunk_index", "The file chunk index is invalid.");
  }
  const nonce = new Uint8Array(new ArrayBuffer(12));
  new DataView(nonce.buffer).setBigUint64(4, BigInt(index), false);
  return nonce;
}

function arrayBuffer(bytes: Uint8Array): ArrayBuffer {
  return Uint8Array.from(bytes).buffer;
}

File: crates/connect-protocol/src/file_crypto.rs
use crate::crypto::{RelayCryptoError, RelayIdentity};
use crate::{
    file_frame_authenticated_data, FileFrame, FileFrameError, FileFrameHeader, FileFrameKind,
    FileTransferDirection, FileTransferProtection, FILE_TRANSFER_PROTOCOL_VERSION,
    RELAY_ENCRYPTION_SUITE,
};
use aes_gcm::aead::{Aead, KeyInit, Payload};
use aes_gcm::{Aes256Gcm, Nonce};
use hkdf::Hkdf;
use sha2::{Digest, Sha256};
use uuid::Uuid;

pub const FILE_TRANSFER_ENCRYPTION_SUITE: &str = RELAY_ENCRYPTION_SUITE;
const FILE_KEY_INFO: &[u8] = b"mdbase-connect file chunk key v1";

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct FileTransferBinding {
    pub grant_id: Uuid,
    pub application_id: Uuid,
    pub connector_id: Uuid,
    pub authority_id: Uuid,
    pub collection_id: Uuid,
    pub scope_epoch: u64,
    pub key_id: String,
    pub transfer_id: Uuid,
    pub direction: FileTransferDirection,
}

impl FileTransferBinding {
    fn validate(&self) -> Result<(), FileCryptoError> {
        if self.scope_epoch == 0
            || self.key_id.is_empty()
            || self.key_id.len() > 200
            || self.key_id.contains('|')
        {
            return Err(FileCryptoError::InvalidBinding);
        }
        Ok(())
    }

    fn context(&self) -> String {
        format!(
            "mdbase-connect|file-transfer|{}|{}|{}|{}|{}|{}|{}|{}|{}|{}|{}",
            FILE_TRANSFER_PROTOCOL_VERSION,
            FILE_TRANSFER_ENCRYPTION_SUITE,
            self.grant_id,
            self.application_id,
            self.connector_id,
            self.authority_id,
            self.collection_id,
            self.scope_epoch,
            self.key_id,
            self.transfer_id,
            self.direction.as_str(),
        )
    }

    fn validate_header(&self, header: &FileFrameHeader) -> Result<(), FileCryptoError> {
        if header.protocol_version != FILE_TRANSFER_PROTOCOL_VERSION
            || header.protection != FileTransferProtection::GrantAeadV1
            || header.grant_id != self.grant_id
            || header.authority_id != self.authority_id
            || header.collection_id != self.collection_id
            || header.transfer_id != self.transfer_id
            || header.direction != self.direction
            || header.scope_epoch != self.scope_epoch
            || header.key_id.as_deref() != Some(self.key_id.as_str())
        {
            return Err(FileCryptoError::HeaderBindingMismatch);
        }
        Ok(())
    }
}

impl FileTransferDirection {
    pub(crate) fn as_str(self) -> &'static str {
        match self {
            Self::Upload => "upload",
            Self::Download => "download",
        }
    }
}

#[derive(Debug, thiserror::Error)]
pub enum FileCryptoError {
    #[error("invalid file transfer encryption binding")]
    InvalidBinding,
    #[error("file frame does not match its transfer encryption binding")]
    HeaderBindingMismatch,
    #[error("file chunk plaintext length does not match its header")]
    PlaintextLengthMismatch,
    #[error("file chunk authentication failed")]
    AuthenticationFailed,
    #[error(transparent)]
    Frame(#[from] FileFrameError),
    #[error(transparent)]
    Relay(#[from] RelayCryptoError),
}

pub struct FileTransferCipher {
    binding: FileTransferBinding,
    key: [u8; 32],
}

impl FileTransferCipher {
    pub fn derive(
        identity: &RelayIdentity,
        peer_public_key: &str,
        binding: FileTransferBinding,
    ) -> Result<Self, FileCryptoError> {
        let shared_secret = identity.shared_secret(peer_public_key)?;
        Self::from_shared_secret(&shared_secret, binding)
    }

    #[doc(hidden)]
    pub fn from_shared_secret(
        shared_secret: &[u8],
        binding: FileTransferBinding,
    ) -> Result<Self, FileCryptoError> {
        binding.validate()?;
        if shared_secret.is_empty() {
            return Err(FileCryptoError::InvalidBinding);
        }
        let context = binding.context();
        let salt = Sha256::digest(context.as_bytes());
        let hkdf = Hkdf::<Sha256>::new(Some(&salt), shared_secret);
        let mut key = [0; 32];
        hkdf.expand(FILE_KEY_INFO, &mut key)
            .map_err(|_| FileCryptoError::InvalidBinding)?;
        Ok(Self { binding, key })
    }

    pub fn encrypt_chunk(
        &self,
        kind: FileFrameKind,
        header: FileFrameHeader,
        plaintext: &[u8],
    ) -> Result<FileFrame, FileCryptoError> {
        self.binding.validate_header(&header)?;
        if plaintext.len() != header.plaintext_length as usize {
            return Err(FileCryptoError::PlaintextLengthMismatch);
        }
        let authenticated_data = file_frame_authenticated_data(kind, &header)?;
        let cipher =
            Aes256Gcm::new_from_slice(&self.key).map_err(|_| FileCryptoError::InvalidBinding)?;
        let nonce_bytes = chunk_nonce(header.chunk_index);
        let payload = cipher
            .encrypt(
                &Nonce::from(nonce_bytes),
                Payload {
                    msg: plaintext,
                    aad: &authenticated_data,
                },
            )
            .map_err(|_| FileCryptoError::AuthenticationFailed)?;
        Ok(FileFrame {
            kind,
            header,
            payload,
        })
    }

    pub fn decrypt_chunk(&self, frame: &FileFrame) -> Result<Vec<u8>, FileCryptoError> {
        self.binding.validate_header(&frame.header)?;
        let authenticated_data = file_frame_authenticated_data(frame.kind, &frame.header)?;
        let cipher =
            Aes256Gcm::new_from_slice(&self.key).map_err(|_| FileCryptoError::InvalidBinding)?;
        let nonce_bytes = chunk_nonce(frame.header.chunk_index);
        cipher
            .decrypt(
                &Nonce::from(nonce_bytes),
                Payload {
                    msg: &frame.payload,
                    aad: &authenticated_data,
                },
            )
            .map_err(|_| FileCryptoError::AuthenticationFailed)
    }
}

fn chunk_nonce(chunk_index: u64) -> [u8; 12] {
    let mut nonce = [0; 12];
    nonce[4..].copy_from_slice(&chunk_index.to_be_bytes());
    nonce
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{DEFAULT_FILE_CHUNK_BYTES, FILE_TRANSFER_PROTOCOL_VERSION};
    use base64::{engine::general_purpose::STANDARD as BASE64, Engine as _};

    fn binding(direction: FileTransferDirection) -> FileTransferBinding {
        FileTransferBinding {
            grant_id: Uuid::from_u128(1),
            application_id: Uuid::from_u128(2),
            connector_id: Uuid::from_u128(3),
            authority_id: Uuid::from_u128(4),
            collection_id: Uuid::from_u128(5),
            scope_epoch: 7,
            key_id: "key-3".to_string(),
            transfer_id: Uuid::from_u128(6),
            direction,
        }
    }

    fn header(direction: FileTransferDirection) -> FileFrameHeader {
        let binding = binding(direction);
        FileFrameHeader {
            protocol_version: FILE_TRANSFER_PROTOCOL_VERSION,
            protection: FileTransferProtection::GrantAeadV1,
            grant_id: binding.grant_id,
            authority_id: binding.authority_id,
            collection_id: binding.collection_id,
            transfer_id: binding.transfer_id,
            direction,
            chunk_size: DEFAULT_FILE_CHUNK_BYTES,
            chunk_index: 0,
            offset: 0,
            plaintext_length: 12,
            total_size: 12,
            scope_epoch: binding.scope_epoch,
            key_id: Some(binding.key_id),
        }
    }

    #[test]
    fn transfer_keys_interoperate_and_are_direction_separated() {
        let application = RelayIdentity::generate();
        let connector = RelayIdentity::generate();
        let upload = binding(FileTransferDirection::Upload);
        let app_cipher =
            FileTransferCipher::derive(&application, &connector.public_key(), upload.clone())
                .unwrap();
        let connector_cipher =
            FileTransferCipher::derive(&connector, &application.public_key(), upload).unwrap();
        let frame = app_cipher
            .encrypt_chunk(
                FileFrameKind::UploadChunk,
                header(FileTransferDirection::Upload),
                b"hello binary",
            )
            .unwrap();
        assert_eq!(
            connector_cipher.decrypt_chunk(&frame).unwrap(),
            b"hello binary"
        );

        let download = FileTransferCipher::derive(
            &connector,
            &application.public_key(),
            binding(FileTransferDirection::Download),
        )
        .unwrap();
        assert!(download.decrypt_chunk(&frame).is_err());
    }

    #[test]
    fn every_bound_header_field_is_authenticated() {
        let cipher = FileTransferCipher::from_shared_secret(
            b"fixed shared secret for tests",
            binding(FileTransferDirection::Upload),
        )
        .unwrap();
        let frame = cipher
            .encrypt_chunk(
                FileFrameKind::UploadChunk,
                header(FileTransferDirection::Upload),
                b"hello binary",
            )
            .unwrap();

        let mut tampered = frame.clone();
        tampered.header.total_size += 1;
        assert!(matches!(
            cipher.decrypt_chunk(&tampered),
            Err(FileCryptoError::AuthenticationFailed)
        ));
        let mut tampered = frame.clone();
        tampered.payload[0] ^= 1;
        assert!(matches!(
            cipher.decrypt_chunk(&tampered),
            Err(FileCryptoError::AuthenticationFailed)
        ));
    }

    #[test]
    fn file_crypto_v1_matches_the_shared_browser_fixture() {
        let fixture: serde_json::Value = serde_json::from_str(include_str!(
            "../../../packages/protocol/test/fixtures/file-crypto-v1.json"
        ))
        .unwrap();
        let binding = FileTransferBinding {
            grant_id: fixture["binding"]["grantId"]
                .as_str()
                .unwrap()
                .parse()
                .unwrap(),
            application_id: fixture["binding"]["applicationId"]
                .as_str()
                .unwrap()
                .parse()
                .unwrap(),
            connector_id: fixture["binding"]["encryption"]["connector_id"]
                .as_str()
                .unwrap()
                .parse()
                .unwrap(),
            authority_id: fixture["binding"]["authorityId"]
                .as_str()
                .unwrap()
                .parse()
                .unwrap(),
            collection_id: fixture["binding"]["encryption"]["collection_id"]
                .as_str()
                .unwrap()
                .parse()
                .unwrap(),
            scope_epoch: fixture["binding"]["encryption"]["scope_epoch"]
                .as_u64()
                .unwrap(),
            key_id: fixture["binding"]["encryption"]["key_id"]
                .as_str()
                .unwrap()
                .to_string(),
            transfer_id: fixture["binding"]["transferId"]
                .as_str()
                .unwrap()
                .parse()
                .unwrap(),
            direction: FileTransferDirection::Upload,
        };
        let shared_secret = BASE64
            .decode(fixture["shared_secret_base64"].as_str().unwrap())
            .unwrap();
        let plaintext = BASE64
            .decode(fixture["plaintext_base64"].as_str().unwrap())
            .unwrap();
        let header = serde_json::from_value(fixture["header"].clone()).unwrap();
        let cipher = FileTransferCipher::from_shared_secret(&shared_secret, binding).unwrap();
        let frame = cipher
            .encrypt_chunk(FileFrameKind::UploadChunk, header, &plaintext)
            .unwrap();
        assert_eq!(
            BASE64.encode(frame.encode().unwrap()),
            fixture["frame_base64"].as_str().unwrap()
        );
        assert_eq!(cipher.decrypt_chunk(&frame).unwrap(), plaintext);
    }
}
